tee_supplicant/src/tee_supplicant.c

```diff
@@ -120,6 +131,462 @@ static int load_ta(uint32_t num_params, struct tee_param *params)
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if (close((int)params[0].b) < 0) {
+		int rc = -errno;
+		LOG_ERR("failed to close file (%d)", rc);
+		return TEEC_ERROR_GENERIC;
+	}
+
+	return TEEC_SUCCESS;
+}
+
+static int tee_fs_read(size_t num_params, struct tee_param *params)
+{
+	int fd;
+	off_t offset, position;
+	size_t len, sz;
+	struct tee_shm *shm;
+	void *buf;
+
+	if (num_params != 2) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT ||
+	    (params[1].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_MEMREF_OUTPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	fd = params[0].b;
+	offset = params[0].c;
+
+	/*TODO do all safety checks */
+	shm = (struct tee_shm *)params[1].c;
+	len = params[1].b;
+
+	if (!shm || !shm->addr) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	buf = shm->addr;
+	position = lseek(fd, offset, SEEK_SET);
+	if ((int)position < 0) {
+		int rc = -errno;
+		LOG_ERR("invalid offset %lu (%d)", offset, rc);
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	/*TODO: handle sz < len */
+	sz = read(fd, buf, len);
+	if ((int)sz < 0) {
+		int rc = -errno;
+		LOG_ERR("read failure (%d)", rc);
+		return TEEC_ERROR_GENERIC;
+	}
+
+	params[1].b = sz;
+	return TEEC_SUCCESS;
+}
+
+static int tee_fs_write(size_t num_params, struct tee_param *params)
+{
+	int fd;
+	off_t offset, position;
+	size_t len, sz;
+	struct tee_shm *shm;
+	void *buf;
+
+	if (num_params != 2) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT ||
+	    (params[1].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_MEMREF_INPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	fd = params[0].b;
+	offset = params[0].c;
+
+	/*TODO do all safety checks */
+	shm = (struct tee_shm *)params[1].c;
+	len = params[1].b;
+
+	if (!shm || !shm->addr) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	buf = shm->addr;
+	position = lseek(fd, offset, SEEK_SET);
+	if ((int)position < 0) {
+		int rc = -errno;
+		LOG_ERR("invalid offset %lu (%d)", offset, rc);
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	sz = write(fd, buf, len);
+	if ((int)sz < 0) {
+		/*TODO: handle error cases */
+		int rc = -errno;
+		LOG_ERR("write failure (%d)", rc);
+		return TEEC_ERROR_GENERIC;
+	}
+
+	params[1].b = sz;
+	return TEEC_SUCCESS;
+}
+
+static int tee_fs_truncate(size_t num_params, struct tee_param *params)
+{
+	int rc, fd;
+	off_t len;
+	void *ptr = NULL;
+	struct fs_file_t *file;
+
+	if (num_params != 1 || (params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	fd = params[0].b;
+	len = params[0].c;
+	ptr = z_get_fd_obj(fd, NULL, 0);
+
+	if (!ptr) {
+		LOG_ERR("descriptor %d not found", fd);
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	/*HACK: this ptr is of hidden posix_fs_desc type, begins w/ file attr */
```

Comment:
You are doing this because zephyr POSIX layer does not support `truncate`?

Reply:
yes, that's right